File: frontend/tests/mocks/scenarios/permissions.ts
// path: tests/mocks/scenarios/permissions.ts
// version: 3 - Converted to MSW v1 syntax

import { rest } from 'msw'
import { mockUsers } from '../data/users'
import { getServer } from '../server-instance'

// Current permission level
let currentPermissionLevel: 'user' | 'manager' | 'root' = 'user'

/**
 * Set current permission level
 */
export function setPermissionLevel(level: 'user' | 'manager' | 'root') {
  currentPermissionLevel = level
}

/**
 * Get current permission level
 */
export function getPermissionLevel(): 'user' | 'manager' | 'root' {
  return currentPermissionLevel
}

/**
 * Reset permission level to default
 */
export function resetPermissionLevel() {
  currentPermissionLevel = 'user'
}

/**
 * Apply permission scenario helper (for handler use)
 */
export function applyPermissionScenario(resource: string, action: string): any {
  // Returns null if permission granted, error response if denied
  const deniedResources: Record<string, string[]> = {
    user: ['users', 'user-groups', 'admin'],
    manager: ['admin', 'user-groups-create'],
  }

  const denied = deniedResources[currentPermissionLevel] || []
  
  if (denied.includes(resource) || denied.includes(`${resource}-${action}`)) {
    return {
      status: 403,
      body: { error: 'Forbidden - insufficient permissions' },
    }
  }

  return null
}

/**
 * Permission scenarios for testing - MSW v1 syntax
 */
export const permissionScenarios = {
  /**
   * Regular user permissions (limited access)
   */
  userPermissions: () => {
    setPermissionLevel('user')
    getServer().use(
      rest.get('/api/auth/verify', (req, res, ctx) => {
        const user = mockUsers.find(u => u.role === 'user' && u.status === 'active')
        return res(
          ctx.status(200),
          ctx.json({ user })
        )
      }),
      // Block access to user management
      rest.get('/api/users', (req, res, ctx) => {
        return res(
          ctx.status(403),
          ctx.json({ error: 'Forbidden - insufficient permissions' })
        )
      }),
      // Block access to group management
      rest.post('/api/user-groups', (req, res, ctx) => {
        return res(
          ctx.status(403),
          ctx.json({ error: 'Forbidden - insufficient permissions' })
        )
      })
    )
  },

  /**
   * Manager permissions (group management)
   */
  managerPermissions: () => {
    setPermissionLevel('manager')
    getServer().use(
      rest.get('/api/auth/verify', (req, res, ctx) => {
        const manager = mockUsers.find(u => u.role === 'manager')
        return res(
          ctx.status(200),
          ctx.json({ user: manager })
        )
      }),
      // Allow access to users in managed groups
      rest.get('/api/users', (req, res, ctx) => {
        const managedUsers = mockUsers.filter(u => 
          u.groupIds?.some(gid => gid === 'group-engineering')
        )
        return res(
          ctx.status(200),
          ctx.json({ users: managedUsers })
        )
      }),
      // Block creating new groups
      rest.post('/api/user-groups', (req, res, ctx) => {
        return res(
          ctx.status(403),
          ctx.json({ error: 'Forbidden - only root can create groups' })
        )
      })
    )
  },

  /**
   * Root permissions (full access)
   */
  rootPermissions: () => {
    setPermissionLevel('root')
    getServer().use(
      rest.get('/api/auth/verify', (req, res, ctx) => {
        const root = mockUsers.find(u => u.role === 'root')
        return res(
          ctx.status(200),
          ctx.json({ user: root })
        )
      })
      // No restrictions for root - handlers will allow everything
    )
  },

  /**
   * User trying to access admin features
   */
  unauthorizedAdminAccess: () => {
    setPermissionLevel('user')
    getServer().use(
      rest.post('/api/admin/maintenance', (req, res, ctx) => {
        return res(
          ctx.status(403),
          ctx.json({ error: 'Forbidden - admin access required' })
        )
      }),
      rest.post('/api/auth/revoke-all-sessions', (req, res, ctx) => {
        return res(
          ctx.status(403),
          ctx.json({ error: 'Forbidden - admin access required' })
        )
      })
    )
  },

  /**
   * Manager trying to access root-only features
   */
  managerAccessingRootFeatures: () => {
    setPermissionLevel('manager')
    getServer().use(
      rest.post('/api/user-groups', (req, res, ctx) => {
        return res(
          ctx.status(403),
          ctx.json({ error: 'Forbidden - only root can create groups' })
        )
      }),
      rest.put('/api/users/:id/role', (req, res, ctx) => {
        return res(
          ctx.status(403),
          ctx.json({ error: 'Forbidden - only root can change roles' })
        )
      }),
      rest.post('/api/admin/maintenance', (req, res, ctx) => {
        return res(
          ctx.status(403),
          ctx.json({ error: 'Forbidden - only root can toggle maintenance' })
        )
      })
    )
  },
}

/**
 * Check if current user has permission
 */
export function checkPermission(requiredLevel: 'user' | 'manager' | 'root'): boolean {
  const levels = { user: 1, manager: 2, root: 3 }
  return levels[currentPermissionLevel] >= levels[requiredLevel]
}

File: backend/src/core/permissions.py
"""
Path: backend/src/core/permissions.py
Version: 2

Permission utilities for role-based access control
Implements role hierarchy: user < manager < root
"""

from typing import Dict, Any

# Role hierarchy levels
ROLE_LEVELS = {
    "user": 1,
    "manager": 2,
    "root": 3,
}


def check_permission(user: Dict[str, Any], required_role: str) -> bool:
    """
    Check if user has required permission level
    
    Role hierarchy:
        user (level 1) - Basic user access
        manager (level 2) - Can manage users and content
        root (level 3) - Full system access
    
    A user with higher role level can access lower level resources.
    For example, a manager can do everything a user can do.
    
    Args:
        user: User dict with "role" field
        required_role: Required role ("user" | "manager" | "root")
        
    Returns:
        True if user has sufficient permissions, False otherwise
        
    Example:
        user = {"role": "manager"}
        
        check_permission(user, "user")     # True (manager >= user)
        check_permission(user, "manager")  # True (manager == manager)
        check_permission(user, "root")     # False (manager < root)
    """
    user_role = user.get("role", "user")
    
    # Get role levels
    user_level = ROLE_LEVELS.get(user_role, 0)
    required_level = ROLE_LEVELS.get(required_role, 999)
    
    # User must have equal or higher level
    return user_level >= required_level


def has_role(user: Dict[str, Any], role: str) -> bool:
    """
    Check if user has exact role
    
    Args:
        user: User dict with "role" field
        role: Role to check
        
    Returns:
        True if user has exact role
        
    Example:
        user = {"role": "manager"}
        has_role(user, "manager")  # True
        has_role(user, "root")     # False
    """
    return user.get("role") == role


def is_user(user: Dict[str, Any]) -> bool:
    """Check if user has user role (or higher)"""
    return check_permission(user, "user")


def is_manager(user: Dict[str, Any]) -> bool:
    """Check if user has manager role (or higher)"""
    return check_permission(user, "manager")


def is_root(user: Dict[str, Any]) -> bool:
    """Check if user has root role"""
    return check_permission(user, "root")


def get_user_level(user: Dict[str, Any]) -> int:
    """
    Get user's permission level
    
    Returns:
        Permission level (1=user, 2=manager, 3=root, 0=unknown)
    """
    user_role = user.get("role", "user")
    return ROLE_LEVELS.get(user_role, 0)